packages/ckeditor5-paragraph/src/insertparagraphcommand.ts

```diff
@@ -59,20 +59,44 @@ export default class InsertParagraphCommand extends Command {
 
 		model.change( writer => {
 			const paragraph = writer.createElement( 'paragraph' );
+			const allowedParent = model.schema.findAllowedParent( position, paragraph );
+
+			// It could be there's no ancestor limit that would allow paragraph.
+			// In theory, "paragraph" could be disallowed even in the "$root".
+			if ( !allowedParent ) {
+				return;
+			}
 
 			if ( attributes ) {
 				model.schema.setAllowedAttributes( paragraph, attributes, writer );
 			}
 
-			if ( !model.schema.checkChild( position.parent as Element, paragraph ) ) {
-				const allowedParent = model.schema.findAllowedParent( position, paragraph );
+			if ( position.path.length < 2 ) {
+				model.insertContent( paragraph, position );
+				writer.setSelection( paragraph, 'in' );
+
+				return;
+			}
+
+			// E.g.
+			// <paragraph>[]</paragraph> ---> <paragraph></paragraph><paragraph>[]</paragraph>
+			const isInEmptyBlock = position.isAtStart && position.isAtEnd;
+
+			// E.g.
+			// <paragraph>foo[]</paragraph> ---> <paragraph>foo</paragraph><paragraph>[]</paragraph>
+			const isAtEndOfTextBlock = position.isAtEnd && position.nodeBefore?.is( '$text' );
+
+			// E.g.
+			// <paragraph>[]foo</paragraph> ---> <paragraph>[]</paragraph><paragraph>foo</paragraph>
+			const isAtStartOfTextBlock = position.isAtStart && position.nodeAfter?.is( '$text' );
```

Comment:
What about inline elements? For example inline image? Maybe you could check `position.parent.isEmpty` instead of checking `nodeBefore`/`nodeAfter`?

Reply:
We decided to keep previous behaviour for **non-text** elements, look at this test: https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-widget/tests/widgettypearound/widgettypearound.js#L1056-L1089
here this kind behaviour is expected for **widgets** then we fix issue only for **text** nodes

Reply:
I'm talking about inline widgets, not block widgets. They are allowed everywhere text is allowed.